http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -98,25 +113,204 @@ class ClusterHelper(dbRef: DbReference,
+        initScripts = initScriptResources.map(resource => GcsPath(initBucketName, GcsObjectName(resource.value)))
+        credentialsFileName = cluster.serviceAccountInfo.notebookServiceAccount.map(_ => s"/etc/${ClusterInitValues.serviceAccountCredentialsFilename}")
+
+        // decide whether to use VPC network
+        lookupProjectLabels = dataprocConfig.projectVPCNetworkLabel.isDefined || dataprocConfig.projectVPCSubnetLabel.isDefined
+        projectLabels <- if (lookupProjectLabels) IO.fromFuture(IO(googleProjectDAO.getLabels(cluster.googleProject.value))) else IO(Map.empty[String, String])
+        clusterVPCSettings = getClusterVPCSettings(projectLabels)
+
+        // Create the cluster
+        createClusterConfig = CreateClusterConfig(machineConfig, initScripts, cluster.serviceAccountInfo.clusterServiceAccount, credentialsFileName, stagingBucketName, cluster.scopes, clusterVPCSettings, cluster.properties, dataprocConfig.customDataprocImage)
+        retryResult <- IO.fromFuture(IO(retryExponentially(whenGoogleZoneCapacityIssue, "Cluster creation failed because zone with adequate resources was not found") { () =>
+          gdDAO.createCluster(cluster.googleProject, cluster.clusterName, createClusterConfig)
+        }))
+
+        operation <- retryResult match {
+          case Right((errors, op)) if errors == List.empty => IO(op)
+          case Right((errors, op)) =>
+            Metrics.newRelic.incrementCounterIO("zoneCapacityClusterCreationFailure", errors.length)
+              .as(op)
+          case Left(errors) =>
+            Metrics.newRelic.incrementCounterIO("zoneCapacityClusterCreationFailure", errors.filter(whenGoogleZoneCapacityIssue).length)
+              .flatMap(_ => IO.raiseError(errors.head))
+        }
+
+        finalCluster = Cluster.createFinal(cluster, operation, stagingBucketName)
+
+      } yield (finalCluster, initBucketName, serviceAccountKeyOpt)
+
+      ioResult.handleErrorWith { throwable =>
+        cleanUpGoogleResourcesOnError(cluster.googleProject,
+          cluster.clusterName, initBucketName,
+          cluster.serviceAccountInfo, serviceAccountKeyOpt) >> IO.raiseError(throwable)
+      }
+    }
+  }
+
+  private def cleanUpGoogleResourcesOnError(googleProject: GoogleProject, clusterName: ClusterName, initBucketName: GcsBucketName, serviceAccountInfo: ServiceAccountInfo, serviceAccountKeyOpt: Option[ServiceAccountKey]): IO[Unit] = {
+    logger.error(s"Cluster creation failed in Google for $googleProject / ${clusterName.value}. Cleaning up resources in Google...")
+
+    // Clean up resources in Google
+    val deleteBucket = bucketHelper.deleteInitBucket(initBucketName) map { _ =>
+      logger.info(s"Successfully deleted init bucket ${initBucketName.value} for ${googleProject.value} / ${clusterName.value}")
+    } recover { case e =>
+      logger.error(s"Failed to delete init bucket ${initBucketName.value} for ${googleProject.value} / ${clusterName.value}", e)
```

Comment:
logging is side effecting. It's better to do `recoverWith{case e => IO(logger.error(xxx))}`  or require `implicit log: Logger[IO]` implicit, and do `recoverWith{case e => log.error(xxx)}`